### PR TITLE
Feat: allow select resource by resource type for gc policy

### PR DESCRIPTION
**Pull request by JooKS-me**
Signed-off-by: zhukunshuai <jookunshuai@gmail.com>

### Description of your changes

About #3395 

/cc @Somefive

### CODE REVIEW

**Comment by Somefive** [comment]
Great work! It is a bit ambiguous to use `resourceType`. It is like `Deployment` or `Service`. How about `OAMRourceType`?

**Comment by JooKS-me** [comment]
> Great work! It is a bit ambiguous to use `resourceType`. It is like `Deployment` or `Service`. How about `OAMRourceType`?

Get it.

File: apis/core.oam.dev/v1alpha1/garbagecollect_types.go [hunk 1]
```diff
@@ -45,12 +45,13 @@ type GarbageCollectPolicyRule struct {
 }
 
 // GarbageCollectPolicyRuleSelector select the targets of the rule
-// if both traitTypes and componentTypes are specified, combination logic is OR
+// if both traitTypes, resourceTypes and componentTypes are specified, combination logic is OR
```
Comment (wonderflow) [inline apis/core.oam.dev/v1alpha1/garbagecollect_types.go:48]:
```suggestion
// if both traitTypes, oamTypes and componentTypes are specified, combination logic is OR
```

File: apis/core.oam.dev/v1alpha1/garbagecollect_types.go [hunk 2]
```diff
@@ -45,12 +45,13 @@ type GarbageCollectPolicyRule struct {
...
 // if one resource is specified with conflict strategies, strategy as component go first.
 type GarbageCollectPolicyRuleSelector struct {
-	CompNames  []string `json:"componentNames"`
-	CompTypes  []string `json:"componentTypes"`
-	TraitTypes []string `json:"traitTypes"`
+	CompNames        []string `json:"componentNames"`
+	CompTypes        []string `json:"componentTypes"`
+	OAMResourceTypes []string `json:"OAMResourceTypes"`
```
Comment (wonderflow) [inline apis/core.oam.dev/v1alpha1/garbagecollect_types.go:53]:
```suggestion
	OAMResourceTypes []string `json:"oamTypes"`
```

File: docs/examples/app-with-policy/gc-policy/persist-resources.md [hunk 3]
```diff
@@ -124,3 +130,74 @@ spec:
...
+        - type: expose
+          properties:
+            port: [8000]
+  policies:
+    - name: garbage-collect
+      type: garbage-collect
+      properties:
+        rules:
+          - selector:
+              resourceTypes:
```
Comment (wonderflow) [inline docs/examples/app-with-policy/gc-policy/persist-resources.md:160]:
```suggestion
              oamTypes:
```

File: docs/examples/app-with-policy/gc-policy/persist-resources.md [hunk 4]
```diff
@@ -124,3 +130,74 @@ spec:
...
```
Comment (wonderflow) [inline docs/examples/app-with-policy/gc-policy/persist-resources.md:190]:
```suggestion
              oamTypes:
```

**Review by wonderflow — CHANGES_REQUESTED** [review]
please align the example doc, annotation and code.

**Comment by JooKS-me** [comment]
> please align the example doc, annotation and code.

ok

**Pull request merged**